### PR TITLE
internal/{charmstore,v5}: fix list results

### PR DESCRIPTION
**Pull request by rogpeppe**
When some list results had different ACLs from others,
we filtered out all but the most recent revision, so
we could fail to show any results even though some
results could be shown.

We fix that by changing ListQuery so that it returns
all matching charms that are in different channels;
then we filter those results after retrieving from MongoDB.

This changes the semantics of sorting by owner because
sorting is now done based on the returned results which
don't necessarily contain the actual owner, so we sort by
the User field in the returned URL, which means that
promulgated charms will always be returned first in that case.

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2174/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2175/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2176/

File: internal/v5/list_test.go [hunk 1]
```diff
@@ -7,13 +7,13 @@ import (
...
 	"encoding/json"
 	"net/http"
-	"sort"
 	"strings"
 
 	"github.com/juju/idmclient"
 	"github.com/juju/loggo"
 	jc "github.com/juju/testing/checkers"
 	"github.com/juju/testing/httptesting"
+	"github.com/kr/pretty"
```
Comment (mhilton) [inline internal/v5/list_test.go]: "that shouldn't be there" - @rogpeppe 2018

**Review by mhilton — APPROVED** [review]
LGTM

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2177/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2178/

**Review by fabricematrat — APPROVED** [review]

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**